code/modules/mining/machinery/material_extractor.dm

```diff
@@ -0,0 +1,392 @@
+// /obj/machinery/atmospherics/unary/material/extractor/proc/force_process_reagents()
+// 	atom_flags &= ~(ATOM_FLAG_NO_TEMP_CHANGE | ATOM_FLAG_NO_REACT)
+// 	. = reagents.process_reactions()
+// 	atom_flags |= ATOM_FLAG_NO_TEMP_CHANGE | ATOM_FLAG_NO_REACT
+
+/obj/machinery/atmospherics/unary/material/extractor/proc/extract(var/list/obj/L)
+	//No point processing anything if the processing buffer is full
+	if(is_buffer_full())
+		log_debug("EXTRACTING : BUFFER FULL")
+		return 
+
+	for(var/obj/O in L)
+		var/free_volume = round(REAGENTS_FREE_SPACE(reagents), GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+		if(free_volume <= 0)
+			break
+		for(var/k in O.matter)
+			free_volume = round(REAGENTS_FREE_SPACE(reagents), GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+			//Don't process over our internal buffer
+			if(free_volume <= 0)
+				break
+			var/matter_units_removed = round(min(O.matter[k], free_volume / REAGENT_UNITS_PER_MATERIAL_UNIT), GAS_EXTRACTOR_MIN_REAGENT_AMOUNT) //Removed only what can be stored
+			var/reagent_units = round(matter_units_removed * REAGENT_UNITS_PER_MATERIAL_UNIT, GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+			reagents.add_reagent(k, reagent_units)
+			log_debug("EXTRACTING : added [reagent_units] units of [k] to internal tank with [free_volume] units of free space")
+
+			//Since we might have to extract over several ticks, if our internal tank is full, we just subtract it from the matter list of the ore
+			O.matter[k] -= matter_units_removed
+			if(round(O.matter[k], 0.1) <= 0)
+				O.matter -= k
+
+		if(length(O.matter) < 1)
+			qdel(O)	//Clean up any ores empty of matter
+
+/obj/machinery/atmospherics/unary/material/extractor/proc/dump_result()
+	//Only dump when everything in the tank is processed
+	if(SSmaterials.active_holders[reagents])
+		return
+
+	for(var/mat in reagents?.reagent_volumes)
+		var/decl/material/M = GET_DECL(mat)
+		var/available_volume = round(REAGENT_VOLUME(reagents, M.type), GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+		
+		//Don't bother if we got a really small quatity, and just get rid of it
+		if(available_volume < GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+			reagents.clear_reagent(M.type)
+			continue
+
+		switch(M.phase_at_temperature(temperature))
+			if(MAT_PHASE_SOLID)
+				dump_solid(M, available_volume)
+			if(MAT_PHASE_LIQUID)
+				dump_liquid(M, available_volume)
+			if(MAT_PHASE_GAS)
+				dump_gas(M, available_volume)
+			if(MAT_PHASE_PLASMA)
+				dump_gas(M, available_volume)
+
+/obj/machinery/atmospherics/unary/material/extractor/proc/dump_solid(var/decl/material/M, var/available_volume)
+	//Check if we got enough to dump out a single sheet at least
+	if(available_volume < REAGENT_UNITS_PER_MATERIAL_SHEET)
+		return
+	var/expected_sheets = round(available_volume / REAGENT_UNITS_PER_MATERIAL_SHEET)
+	var/removed_volume = round(expected_sheets * REAGENT_UNITS_PER_MATERIAL_SHEET, GAS_EXTRACTOR_MIN_REAGENT_AMOUNT)
+	M.place_sheet(get_output_loc(), expected_sheets)
```

Comment:
This got changed to `create_object()`, same arg order though.

Reply:
Yeah, fixed that